From dce47396d64e417e6782927bb5569987d4b5ebc3 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Sun, 17 Sep 2017 12:31:10 +0300
Subject: [PATCH 07/18] Add support for Samsung extended AGPS

The AGPS implementation in the GPS chipset used in some Samsung devices
(i9100, i9300) can make use of the Psc field. Adapt the relevant
functions hiding the changes incompatible with other devices under the
AGPS_USE_PSC
define.

Credit to Qaweck from xda-developers forum for finding the meaning
of the field.

Change-Id: I03895b503166e23de9aebb2bea0df067dfa171ef
---
 .../java/com/android/server/location/GnssLocationProvider.java   | 4 ++--
 .../jni/com_android_server_location_GnssLocationProvider.cpp     | 9 +++++++--
 2 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/services/core/java/com/android/server/location/GnssLocationProvider.java b/services/core/java/com/android/server/location/GnssLocationProvider.java
index cd3c05f..eb2844f 100644
--- a/services/core/java/com/android/server/location/GnssLocationProvider.java
+++ b/services/core/java/com/android/server/location/GnssLocationProvider.java
@@ -2082,7 +2082,7 @@ public class GnssLocationProvider implements LocationProviderInterface {
                     type = AGPS_REF_LOCATION_TYPE_GSM_CELLID;
                 }
                 native_agps_set_ref_location_cellid(type, mcc, mnc,
-                        gsm_cell.getLac(), gsm_cell.getCid());
+                        gsm_cell.getLac(), gsm_cell.getCid(), gsm_cell.getPsc());
             } else {
                 Log.e(TAG,"Error getting cell location info.");
             }
@@ -2548,7 +2548,7 @@ public class GnssLocationProvider implements LocationProviderInterface {
 
     // AGPS ril suport
     private native void native_agps_set_ref_location_cellid(int type, int mcc, int mnc,
-            int lac, int cid);
+            int lac, int cid, int psc);
     private native void native_agps_set_id(int type, String setid);
 
     private native void native_update_network_state(boolean connected, int type,
diff --git a/services/core/jni/com_android_server_location_GnssLocationProvider.cpp b/services/core/jni/com_android_server_location_GnssLocationProvider.cpp
index edd7d53..02f0979 100644
--- a/services/core/jni/com_android_server_location_GnssLocationProvider.cpp
+++ b/services/core/jni/com_android_server_location_GnssLocationProvider.cpp
@@ -1316,7 +1316,7 @@ static jint android_location_GnssLocationProvider_read_sv_status(JNIEnv* env, jo
 }
 
 static void android_location_GnssLocationProvider_agps_set_reference_location_cellid(
-        JNIEnv* /* env */, jobject /* obj */, jint type, jint mcc, jint mnc, jint lac, jint cid) {
+        JNIEnv* /* env */, jobject /* obj */, jint type, jint mcc, jint mnc, jint lac, jint cid, jint psc) {
     IAGnssRil::AGnssRefLocation location;
 
     if (agnssRilIface == nullptr) {
@@ -1331,6 +1331,11 @@ static void android_location_GnssLocationProvider_agps_set_reference_location_ce
           location.cellID.mcc = mcc;
           location.cellID.mnc = mnc;
           location.cellID.lac = lac;
+#ifdef AGPS_USE_PSC
+          location.cellID.psc = psc;
+#else
+          (void)psc;
+#endif
           location.cellID.cid = cid;
           break;
         default:
@@ -1935,7 +1940,7 @@ static const JNINativeMethod sMethods[] = {
             "(ILjava/lang/String;)V",
             reinterpret_cast<void *>(android_location_GnssLocationProvider_agps_set_id)},
     {"native_agps_set_ref_location_cellid",
-            "(IIIII)V",
+            "(IIIIII)V",
             reinterpret_cast<void *>(
                     android_location_GnssLocationProvider_agps_set_reference_location_cellid)},
     {"native_set_agps_server",
-- 
2.7.4

